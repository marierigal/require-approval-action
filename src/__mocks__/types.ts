import { REVIEW_STATE } from '../constants';
import {
  DeleteWorkflowRunParams,
  ListReviewsParams,
  ListWorkflowRunsParams,
} from '../types';

export interface MockCore {
  __setMockInputs: (inputs: Record<string, string>) => void;
  getInput: (name: string) => string | undefined;
}

export interface MockGithub {
  __setMockContext: (pull_number?: number) => MockContext;
  __makeMockReview: (state: REVIEW_STATE, userId: number) => MockReview;
  __setMockReviews: (reviews: MockReview[]) => void;
  __makeMockRun: (runId: number, pull_number: number) => MockRun;
  __setMockRuns: (runs: MockRun[]) => void;
  context: MockContext;
  getOctokit: (token: string) => MockOctokit;
}

export interface MockOctokit {
  pulls: {
    listReviews: (params: ListReviewsParams) => Promise<{ data: MockReview[] }>;
  };
  actions: {
    listWorkflowRuns: (
      params: ListWorkflowRunsParams
    ) => Promise<{ data: { workflow_runs: MockRun[] } }>;
    deleteWorkflowRun: (params: DeleteWorkflowRunParams) => Promise<void>;
  };
}

export interface MockContext {
  eventName: string;
  workflow_id: number;
  repo: {
    owner: string;
    repo: string;
  };
  payload: {
    pull_request?: { number: number };
  };
}

export interface MockReview {
  user: { id: number };
  state: REVIEW_STATE;
}

export interface MockRun {
  id: number;
  pull_requests: { number: number }[];
}
