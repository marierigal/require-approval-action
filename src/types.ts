import { RestEndpointMethodTypes } from '@octokit/rest';

export type ListReviewsParams = RestEndpointMethodTypes['pulls']['listReviews']['parameters'];
export type Review = RestEndpointMethodTypes['pulls']['getReview']['response']['data'];
export type SortedReviews = { approved: Review[]; changesRequested: Review[] };

export type ListWorkflowRunsParams = RestEndpointMethodTypes['actions']['listWorkflowRuns']['parameters'];
export type DeleteWorkflowRunParams = RestEndpointMethodTypes['actions']['deleteWorkflowRun']['parameters'];
export type Run = RestEndpointMethodTypes['actions']['getWorkflowRun']['response']['data'];
