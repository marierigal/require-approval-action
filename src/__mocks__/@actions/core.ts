import { MockCore } from '../types';

const core = jest.createMockFromModule<MockCore>('@actions/core');

let mockInputs: Record<string, string> = {
  token: 'token',
  approvals: '3',
};

function __setMockInputs(inputs: Record<string, string>): void {
  mockInputs = inputs;
}

function getInput(name: string): string | undefined {
  return mockInputs[name];
}

core.__setMockInputs = __setMockInputs;
core.getInput = getInput;

module.exports = core;
