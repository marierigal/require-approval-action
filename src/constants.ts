export enum REVIEW_STATE {
  approved = 'APPROVED',
  changesRequested = 'CHANGES_REQUESTED',
  commented = 'COMMENTED',
}
