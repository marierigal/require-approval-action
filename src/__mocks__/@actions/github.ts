import { REVIEW_STATE } from '../../constants';
import {
  DeleteWorkflowRunParams,
  ListReviewsParams,
  ListWorkflowRunsParams,
} from '../../types';
import {
  MockContext,
  MockGithub,
  MockOctokit,
  MockReview,
  MockRun,
} from '../types';

const github = jest.createMockFromModule<MockGithub>('@actions/github');

const context: MockContext = {
  eventName: 'pull_request_review',
  workflow_id: 1,
  repo: {
    owner: 'owner',
    repo: 'repo',
  },
  payload: {},
};
let mockReviews: MockReview[] = [];
let mockRuns: MockRun[] = [];

function __setMockContext(pull_number?: number): MockContext {
  if (pull_number) {
    context.payload = {
      pull_request: { number: pull_number },
    };
  }

  return context;
}

function __makeMockReview(state: REVIEW_STATE, userId: number): MockReview {
  return {
    user: { id: userId },
    state: state,
  };
}

function __setMockReviews(reviews: MockReview[]): void {
  mockReviews = reviews;
}

function __makeMockRun(runId: number, pull_number: number): MockRun {
  return {
    id: runId,
    pull_requests: [{ number: pull_number }],
  };
}

function __setMockRuns(runs: MockRun[]): void {
  mockRuns = runs;
}

function listReviews(
  _params: ListReviewsParams
): Promise<{ data: MockReview[] }> {
  return new Promise(resolve => resolve({ data: mockReviews }));
}

function listWorkflowRuns(
  _params: ListWorkflowRunsParams
): Promise<{ data: { workflow_runs: MockRun[] } }> {
  return new Promise(resolve =>
    resolve({
      data: { workflow_runs: mockRuns },
    })
  );
}

function deleteWorkflowRun({ run_id }: DeleteWorkflowRunParams): Promise<void> {
  return new Promise(resolve => {
    mockRuns = mockRuns.filter(run => run.id !== run_id);
    resolve();
  });
}

function getOctokit(_token: string): MockOctokit {
  return {
    pulls: {
      listReviews,
    },
    actions: {
      listWorkflowRuns,
      deleteWorkflowRun,
    },
  };
}

github.__setMockContext = __setMockContext;
github.__makeMockReview = __makeMockReview;
github.__setMockReviews = __setMockReviews;
github.__makeMockRun = __makeMockRun;
github.__setMockRuns = __setMockRuns;
github.context = context;
github.getOctokit = getOctokit;

module.exports = github;
