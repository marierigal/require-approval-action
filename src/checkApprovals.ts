import { ExitCode, getInput } from '@actions/core';
import { context } from '@actions/github';
import { getLastReviews } from './getLastReviews';
import { removePreviousRuns } from './removePreviousRuns';

export async function checkApprovals() {
  // Get token from input
  const token = getInput('token');
  if (!token) throw new Error('Github token not found');

  // Only run on `pull_request` events
  if (!context.payload.pull_request) return;

  // Get pull number
  const pull_number = context.payload.pull_request.number;

  // Set base octokit params
  const octokitParams = {
    owner: context.repo.owner,
    repo: context.repo.repo,
  };

  // Remove previous runs triggered by the `pull_request` event
  if (context.eventName === 'pull_request_review') {
    await removePreviousRuns(
      token,
      {
        ...octokitParams,
        workflow_id: context.workflow,
        event: 'pull_request',
      },
      pull_number
    );
  }

  // Remove previous runs triggered by the `pull_request_review` event
  if (context.eventName === 'pull_request') {
    await removePreviousRuns(
      token,
      {
        ...octokitParams,
        workflow_id: context.workflow,
        event: 'pull_request_review',
      },
      pull_number
    );
  }

  // Get approvals from input
  const requiredApprovals = parseInt(getInput('approvals'), 10) ?? 1;
  if (isNaN(requiredApprovals))
    throw new Error('Approvals input must be a number');
  if (requiredApprovals < 1) return ExitCode.Success;

  // Get PR reviews
  const lastReviews = await getLastReviews(token, {
    ...octokitParams,
    pull_number,
  });
  const approvedReviews = lastReviews.approved.length;
  const rejectedReviews = lastReviews.changesRequested.length;

  if (rejectedReviews) {
    throw new Error(`This pull request has ${rejectedReviews} request changes`);
  }

  if (requiredApprovals > approvedReviews) {
    throw new Error(`Missing ${requiredApprovals - approvedReviews} approvals`);
  }

  return ExitCode.Success;
}
