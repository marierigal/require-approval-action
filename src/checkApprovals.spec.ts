import * as core from '@actions/core';
import * as github from '@actions/github';
import { MockCore, MockGithub } from './__mocks__/types';

import { checkApprovals } from './checkApprovals';
import { REVIEW_STATE } from './constants';

jest.mock('@actions/core');
jest.mock('@actions/github');

const mockCore = (core as unknown) as MockCore;
const mockGithub = (github as unknown) as MockGithub;

describe('checkApprovals', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should throw an error when the token is not defined', () => {
    mockCore.__setMockInputs({});

    return expect(checkApprovals()).rejects.toMatchSnapshot();
  });

  it('should skip execution when the event is not related to a pull request', () => {
    mockCore.__setMockInputs({ token: 'token' });
    mockGithub.__setMockContext();

    return expect(checkApprovals()).resolves.toBeFalsy();
  });

  it('should fail when requiredApprovals = NaN', () => {
    mockCore.__setMockInputs({ token: 'token', approvals: 'foo' });
    mockGithub.__setMockContext(1);

    return expect(checkApprovals()).rejects.toMatchSnapshot();
  });

  it('should success when requiredApprovals < 1', () => {
    mockCore.__setMockInputs({ token: 'token', approvals: '0' });
    mockGithub.__setMockContext(1);
    mockGithub.__setMockReviews([
      mockGithub.__makeMockReview(REVIEW_STATE.changesRequested, 1),
    ]);

    return expect(checkApprovals()).resolves.toEqual(core.ExitCode.Success);
  });

  it('should fail when rejectedReviews > 0', () => {
    mockCore.__setMockInputs({ token: 'token', approvals: '3' });
    mockGithub.__setMockContext(1);
    mockGithub.__setMockReviews([
      mockGithub.__makeMockReview(REVIEW_STATE.approved, 1),
      mockGithub.__makeMockReview(REVIEW_STATE.changesRequested, 1),
      mockGithub.__makeMockReview(REVIEW_STATE.approved, 2),
    ]);
    mockGithub.__setMockRuns([]);

    return expect(checkApprovals()).rejects.toMatchSnapshot();
  });

  it('should fail when requiredApprovals > approvedReviews', () => {
    mockCore.__setMockInputs({ token: 'token', approvals: '3' });
    mockGithub.__setMockContext(1);
    mockGithub.__setMockReviews([
      mockGithub.__makeMockReview(REVIEW_STATE.approved, 1),
      mockGithub.__makeMockReview(REVIEW_STATE.approved, 1),
      mockGithub.__makeMockReview(REVIEW_STATE.approved, 1),
      mockGithub.__makeMockReview(REVIEW_STATE.commented, 2),
    ]);

    return expect(checkApprovals()).rejects.toMatchSnapshot();
  });

  it('should success when requiredApprovals < approvedReviews', () => {
    mockCore.__setMockInputs({ token: 'token', approvals: '1' });
    mockGithub.__setMockContext(1);
    mockGithub.__setMockReviews([
      mockGithub.__makeMockReview(REVIEW_STATE.approved, 1),
      mockGithub.__makeMockReview(REVIEW_STATE.approved, 2),
    ]);

    return expect(checkApprovals()).resolves.toEqual(core.ExitCode.Success);
  });

  it('should success when requiredApprovals = approvedReviews', () => {
    mockCore.__setMockInputs({ token: 'token', approvals: '2' });
    mockGithub.__setMockContext(1);
    mockGithub.__setMockReviews([
      mockGithub.__makeMockReview(REVIEW_STATE.approved, 1),
      mockGithub.__makeMockReview(REVIEW_STATE.changesRequested, 1),
      mockGithub.__makeMockReview(REVIEW_STATE.approved, 1),
      mockGithub.__makeMockReview(REVIEW_STATE.approved, 2),
    ]);

    return expect(checkApprovals()).resolves.toEqual(core.ExitCode.Success);
  });
});
