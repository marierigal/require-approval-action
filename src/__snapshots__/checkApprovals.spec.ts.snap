// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkApprovals should fail when rejectedReviews > 0 1`] = `[Error: This pull request has 1 request changes]`;

exports[`checkApprovals should fail when requiredApprovals = NaN 1`] = `[Error: Approvals input must be a number]`;

exports[`checkApprovals should fail when requiredApprovals > approvedReviews 1`] = `[Error: Missing 2 approvals]`;

exports[`checkApprovals should throw an error when the token is not defined 1`] = `[Error: Github token not found]`;
